### DISCUSSION TITLE
[XState Model Based Testing]: How to fire / mimic events sent from within the machine?

### DISCUSSION THREAD

**Discussion by dsathyakumar** [statelyai/xstate#2559] [Q&A]
Hello 👋 
I've been checking out examples of MBT and from articles like this:
https://css-tricks.com/model-based-testing-in-react-with-state-machines/ 

It states that to go from State A to B, an event is fired and those events are represented here (via xstate machines):

![image](https://user-images.githubusercontent.com/5446493/130863771-5fc15afe-8d2a-415c-aca9-37e8c2b9bb87.png)

while, in the event configs (for model based testing), they're mentioned as:

![image](https://user-images.githubusercontent.com/5446493/130863828-45e8b243-c391-4a5d-a7a9-0d5ef200a315.png)

i have a state machine with one of the states that look like this:

![image](https://user-images.githubusercontent.com/5446493/130863883-7f33308f-7b02-4db9-b977-e6fb68275966.png)

The **VALIDATION_FAIL**, **VALIDATION_SUCCESS** etc., are events fired via `send` actions to the state machine instance. They're not user triggered events & so cannot be mimicked with `fireEvent` in the model based testing scenario. 

How does one write event configs for these, for model based testing?

CC - @davidkpiano

**Comment by dsathyakumar** [comment]
Some other questions I had were:
- What if moving from state A to B did not have any user related event triggers but was triggered by a `send` action from within the machine ?
- Also what about the cases when moving from state A to B did not have visual elements to check on ?
- Could you point me to a full blown example with `services`, `actions` & `guards` ?

CC @davidkpiano @Andarist

**Comment by davidkpiano** [comment]
> The **VALIDATION_FAIL**, **VALIDATION_SUCCESS** etc., are events fired via `send` actions to the state machine instance. They're not user triggered events & so cannot be mimicked with `fireEvent` in the model based testing scenario.

How do those messages come to be sent, though? Is it through the result of a `fetch(...)` call, or some other mechanism? That's the mechanism that needs to be simulated in your E2E tests, because your app works from those external mechanisms, and not by directly `send`-ing to a machine. The test model machine should be completely unaware that XState is even used in the app.

**Reply by dsathyakumar** [reply]
They are events emitted as a result of running validations on the incoming props.
In most of the model based testing examples, I dont see props being passed to the `render` method. Is that intentional ?

**Reply by dsathyakumar** [reply]
@davidkpiano The xstate test demo doesnt work. Do you have a full blown example on how to deal with guards & send actions & running validations based on incoming props & testing usecases where there is no visual change?
This blog here mentions a deal of excruciating work arounds done to get it working:
https://blog.smartive.ch/3-problems-we-encountered-using-xstate-test-with-testcafe-and-their-solutions-492c237b5f85

Greatly appreciate some pointers on how to get this going.
